opensafely/log_stats.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/python3
+import json
+import re
+from argparse import ArgumentParser
+from datetime import datetime
+from pathlib import Path
+
+DESCRIPTION = "Parse and output cohortextractor-stats logs as JSONL"
+TIMESTAMP_PREFIX_REGEX = r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{9}Z"
+COMPLETED_PREFIX_REGEX = r"^Completed successfully"
+KEY_VALUE_REGEX = r"(?<=\s)([^\s=]+)=(.*?)(?=(?:\s[^\s=]+=|$))"
+ACTION_SUMMARY_REGEX = (
+    r"^(state|docker_image_id|job_id|run_by_user|created_at|completed_at):\s(.*)"
+)
+
+
+def add_arguments(parser):
+    parser.add_argument(
+        "--project-dir",
+        help="Project directory (default: current directory)",
+        type=Path,
+        default=".",
+    )
+    parser.add_argument(
+        "--project-name",
+        help="Project name (default: project-dir)",
+        type=str,
+    )
+    parser.add_argument(
+        "--output-file",
+        help="Output json filename (default: log_stats.json)",
+        type=Path,
+        default="log_stats.json",
+    )
+    return parser
+
+
+def _timestamp_for_honeytail(timestamp, ts_format):
+    honeytail_ts_format = "%Y-%m-%dT%H:%M:%S.000000000Z"
+    dt = datetime.strptime(timestamp, ts_format)
+    return dt.strftime(honeytail_ts_format)
+
+
+def parse_log(project_name, current_action, job_id, current_log):
+    current_log_timestamp = re.match(TIMESTAMP_PREFIX_REGEX, current_log).group()
+    current_log = re.sub(r"\s*\n\s*", " ", current_log).strip()
+    log_params = dict(re.findall(KEY_VALUE_REGEX, current_log))
+    ts_format = "%Y-%m-%dT%H:%M:%S"
+    return {
+        "timestamp": current_log_timestamp,
+        "project": project_name,
+        "job_id": job_id,
+        "action": current_action,
+        **log_params,
+    }
+
+
+def parse_stats_logs(raw_logs, project_name, current_action, job_id):
+    for log in raw_logs:
+        yield parse_log(project_name, current_action, job_id, log)
+
+
+def format_summary_stats(project_name, current_action, summary_stats):
+    start_time = summary_stats.get("created_at")
+    end_time = summary_stats.get("completed_at")
+    ts_format = "%Y-%m-%dT%H:%M:%SZ"
+    start_dt = datetime.strptime(start_time, ts_format)
+
+    if start_time and end_time:
+        elapsed = datetime.strptime(end_time, ts_format) - start_dt
+        summary_stats["elapsed_time_secs"] = elapsed.seconds
+        summary_stats["elapsed_time"] = str(elapsed)
+    return {
+        "timestamp": _timestamp_for_honeytail(start_time, ts_format),
+        "project": project_name,
+        "action": current_action,
+        **{k: v for k, v in summary_stats.items()},
+    }
+
+
+def main(project_dir, output_file, project_name=None):
+    log_dir = project_dir / "metadata"
+    project_name = project_name or project_dir.resolve().name
+
+    log_files = list((log_dir).glob("*.log"))
```

Comment:
I'm guessing the parens around `log_dir` here are left over from when it used to read `(project_dir / "metadata")`?

Reply:
indeed :)